feat(source): detect Lua codeblocks via treesitter injection

benarcher2691 opened the pull request:
## Summary

- When `:source` is given a range in a file like vimdoc, use treesitter to detect if the selected text is inside a Lua codeblock (injection)
- If the range is in a Lua injection, execute as Lua instead of Vimscript
- Falls back to current behavior when no treesitter parser is available

This allows sourcing Lua code from help files without explicitly using `:'<,'>lua`.

## Test plan

- [x] Added tests for Lua codeblock detection in vimdoc
- [x] Added tests for Vimscript codeblock (non-Lua injection)
- [x] Added tests for fallback when no treesitter parser
- [x] All existing `:source` tests pass

Closes #36761

benarcher2691 committed:
feat(source): detect Lua codeblocks via treesitter injection

When `:source` is given a range in a file like vimdoc, use treesitter
to detect if the selected text is inside a Lua codeblock (injection).
If so, execute as Lua instead of Vimscript.

This allows sourcing Lua code from help files without explicitly using
`:'<,'>lua`.

Closes #36761

benarcher2691 committed:
ci: retry flaky screen tests

ribru17 commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1316,6 +1316,25 @@ function vim.deprecate(name, alternative, version, plugin, backtrace)
+    return false
+  end
+  -- Parse to ensure injections are processed
+  parser:parse(true)
```
Should only parse the specified range here, `parse(true)` can be very expensive

ribru17 commented on test/functional/ex_cmds/source_spec.lua:
```diff
@@ -296,6 +296,46 @@ describe(':source', function()
+        Example: >lua
+          vim.g.test_vimdoc_lua = 42
+        <]])
+      feed('dd') -- remove trailing empty line from insert
```
Can this line be removed if we just do `insert([[*test.txt* ...`?

justinmk commented on test/functional/ex_cmds/source_spec.lua:
```diff
@@ -296,6 +296,46 @@ describe(':source', function()
+
+        Example: >vim
+          let g:test_vimdoc_vim = 99
+        <]])
```
this can be added to the above test, then just move the cursor. that also makes the test closer to real-world. and the `setlocal filetype=` test can be done here. so then instead of a `describe()` just one `it()`.

benarcher2691 committed:
refactor(source): move source_is_lua to _core/util

- Move treesitter injection detection to `vim._core.util.source_is_lua()`
- Use `NLUA_EXEC_STATIC` macro in runtime.c
- Consolidate tests into single comprehensive test

brianhuster commented:
I think `:h :source` needs updating

justinmk commented:
I'll update the docs.
